module/zfs/sha256.c

```diff
@@ -44,13 +45,25 @@ void
 abd_checksum_SHA256(abd_t *abd, uint64_t size,
     const void *ctx_template, zio_cksum_t *zcp)
 {
+	int ret;
 	SHA2_CTX ctx;
 	zio_cksum_t tmp;
 
+	if (qat_checksum_use_accel(size)) {
+		uint8_t *buf = abd_borrow_buf_copy(abd, size);
+		ret = qat_checksum(ZIO_CHECKSUM_SHA256, buf, size, &tmp);
+		abd_return_buf(abd, buf, size);
+		if (ret == CPA_STATUS_SUCCESS)
+			goto bswap;
```

Comment:
for SHA512, we should just fall back to software, instead of jumping to bswap?

Reply:
Not sure what you mean here. The goto bswap here is meant to handle a strange implementation detail that ZFS always stores SHA256 checksums in big endian order. For SHA512 we don't do this.

Reply:
If we jump to bswap the software implementation will be skipped for SHA512.

Reply:
We only do the jump in the event that QAT checksumming is enabled and if the QAT implementation worked. Is that correct?

Reply:
Im also a little confused with what you mean by "software implementation of SHA512" I removed the QAT implementation of SHA512 since ZFS actually uses SHA512/256, as you pointed out.

Reply:
Sorry for confusion..  I just relook at the code, it is good. 
I have thought SHA512 shared the same function, I just noticed there is another sha512 function which is not changed and will be always in "software" mode. 
Just ignore this comment. thanks!